lib/active_model/serializer/association.rb

```diff
@@ -9,10 +9,58 @@ class Serializer
     # @example
     #  Association.new(:comments, CommentSummarySerializer)
     #
-    Association = Struct.new(:name, :serializer, :options, :links, :meta) do
+    Association = Struct.new(:name, :serializer, :options) do
+      attr_reader :reflection_options, :object
+
+      def initialize(*)
+        super
+        @reflection_options = options.delete(:new_options)
+      end
+
       # @return [Symbol]
       def key
-        options.fetch(:key, name)
+        @key ||= options.fetch(:key) { name }
+      end
+
+      def links
+        reflection_options.fetch(:links)
+      end
+
+      def meta
+        reflection_options.fetch(:meta)
+      end
+
+      def reflection
+        reflection_options.fetch(:reflection)
+      end
+
+      def to_many?
+        reflection.name == :has_many
+      end
+
+      def belongs_to?
+        reflection.foreign_key_on == :self
+      end
+
+      def polymorphic_type
+        "#{reflection.type.to_s.singularize}_type"
+      end
+
+      def value
+        @value ||= association_value = reflection_options.fetch(:association_value_proc).call
+      end
+
+      def serializer
+        defined?(@reified_serializer) ? @reified_serializer : reify!
+      end
```

Comment:
Hack to work around this being an attribute of the Struct, and current public interface method, and me not wanting to make too many unnecessary changes